tests/CppWinRTAuthoringTests.cpp

```diff
@@ -173,6 +175,71 @@ TEST_CASE("CppWinRTAuthoringTests::InStruct", "[property]")
     REQUIRE(test.Prop2() == 33);
 }
 
+// Simple mock of GetValue & SetValue for testing dependency properties.
+struct mock_dependency_object
+{
+    winrt::Windows::Foundation::IInspectable GetValue(winrt::Windows::UI::Xaml::DependencyProperty const& dp) const
+    {
+        return m_values.at(dp);
+    }
+    void SetValue(winrt::Windows::UI::Xaml::DependencyProperty dp, winrt::Windows::Foundation::IInspectable const& value) const
+    {
+        // We have to do a const_cast because ordinarily SetValue is const---
+        // it's delegated to WinRT, and C++/WinRT marks most of its functions
+        // const.
+        const_cast<mock_dependency_object*>(this)->m_values[dp] = value;
+    }
+
+    std::map<winrt::Windows::UI::Xaml::DependencyProperty, winrt::Windows::Foundation::IInspectable> m_values{};
+};
+
+struct my_dependency_properties : mock_dependency_object
+{
+    WIL_DEFINE_DP(my_dependency_properties, int32_t, MyProperty);
+    WIL_DEFINE_DP_WITH_DEFAULT_VALUE_AND_CALLBACK(my_dependency_properties, int32_t, MyPropertyWithDefault, 42, nullptr);
+    WIL_DEFINE_DP(my_dependency_properties, winrt::hstring, MyStringProperty);
+};
+
+namespace winrt
+{
+    // Fake the xaml_typename specialization
+    template<>
+    inline Windows::UI::Xaml::Interop::TypeName xaml_typename<my_dependency_properties>()
+    {
+        static const Windows::UI::Xaml::Interop::TypeName name{hstring{L"my_dependency_properties"}, Windows::UI::Xaml::Interop::TypeKind::Custom};
+        return name;
+    }
+}
+
+TEST_CASE("CppWinRTAuthoringTests::DependencyProperties", "[property]")
+{
+    // Throws if not ensured (?)
+    auto obj = my_dependency_properties{};
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyProperty(); });
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyProperty(42); });
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyPropertyWithDefault(); });
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyPropertyWithDefault(42); });
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyStringProperty(); });
+    // REQUIRE_FAILFAST_MSG(E_NOTIMPL, [&obj] { obj.MyStringProperty(L"foo"); });
+
+    // Register the dependency property
+    my_dependency_properties::EnsureMyPropertyProperty();
+    my_dependency_properties::EnsureMyPropertyWithDefaultProperty();
+    my_dependency_properties::EnsureMyStringPropertyProperty();
+
+    // Now it should work
+    obj.MyProperty(42);

```

Comment:
What is the behavior of querying a property that has not been set yet?  Undefined behavior?  Should this macro require that the properties be default-constructible or they have to provide a default value if not?

Reply:
when you register the property you give it its default value

Reply:
The macro without a default value passes `nullptr` for the `Windows::UI::Xaml::PropertyMetadata`.  That leaves it ambiguous as to the default value.  We'd have to check the XAML documentation (assuming the behavior here is publicly documented).

Reply:
> Passing null for this parameter is equivalent to passing a new PropertyMetadata instance created by calling [PropertyMetadata.Create](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.propertymetadata.create?view=winrt-26100#windows-ui-xaml-propertymetadata-create(windows-ui-xaml-createdefaultvaluecallback)) with null as the default value parameter.

See https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.dependencyproperty.register?view=winrt-26100#definition

Reply:
So calling get without specifying a default value or setting it first would result in `winrt::unbox_value<T>(nullptr)` getting called.  And that seems to throw E_NOINTERFACE. https://github.com/microsoft/cppwinrt/blob/2e4bdb05da1fe86080c523a2e00f81d62f209b17/strings/base_reference_produce.h#L430

Reply:
https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.dependencyproperty.register?view=winrt-26100
> Passing null for this parameter is equivalent to passing a new PropertyMetadata instance created by calling [PropertyMetadata.Create](https://learn.microsoft.com/en-us/uwp/api/windows.ui.xaml.propertymetadata.create?view=winrt-26100#windows-ui-xaml-propertymetadata-create(windows-ui-xaml-createdefaultvaluecallback)) with null as the default value parameter.

This means that for ref types, you get a null default value. For value types, unbox_value will throw `hresult_no_interface` until a value is set (because you'll try to unbox nullptr).

Reply:
Oops, those comments hadn't showed up on my end when I was writing this. @dmachaj is correct, with the addition that it wouldn't throw with ref types, but just return nullptr.